File: scraper4/scrape-albums/scrape_albums/scrape_albums.py
from loguru import logger
from data import delete_tag, pick_next_tag_to_scrape, save_tag_scrape_success
from tag_blacklist import is_tag_blacklisted
from telegram_logger import log_to_telegram

from .remove_tag_duplicates import remove_tag_duplicates

@logger.catch
def scrape_albums():
    logger.debug('scrape_albums()')
    tag = pick_next_tag_to_scrape()
    if tag is None:
        logger.error('Failed to find a tag to scrape albums by')
        return
    if is_tag_blacklisted(tag):
        delete_tag(tag)
        scrape_albums()
        return
    removed_duplicates = remove_tag_duplicates(tag['name'])
    if tag['name'] in removed_duplicates:
        logger.warning(f'{tag["name"]} - removed as a duplicate')
        scrape_albums()
        return
    scrape_albums_by_tag(tag)
    save_tag_scrape_success(tag['name'])
    logger.debug(f'scrapeAlbums: {tag["name"]} - success')
    log_to_telegram(fr'\\#scrape\\_albums\nУспішно зібрано альбоми для тега *{tag.name}*')

File: scraper/src/scrape-albums/scrape-albums.ts
import SQL from '@nearform/sql';
import _ from 'lodash';

import database from '../common/database/index.js';
import { deleteTag } from '../common/database/tag.js';
import isTagBlacklisted from '../common/is-tag-blacklisted.js';
import logToTelegram, {
  escapeTelegramMessage,
} from '../common/log-to-telegram.js';
import logger from '../common/logger.js';
import removeTagDuplicates from '../common/remove-tag-duplicates.js';

import pickTag from './pick-tag.js';
import scrapeAlbumsByTag from './scrape-albums-by-tag.js';

const { includes } = _;

export default async function scrapeAlbums(): Promise<void> {
  logger.debug('scrapeAlbums()');
  const tag = await pickTag();

  if (!tag) {
    logger.error('Failed to find a tag to scrape albums by');
    return;
  }
  if (isTagBlacklisted(tag.name)) {
    await deleteTag(tag);
    await scrapeAlbums();
    return;
  }

  const removedDuplicates = await removeTagDuplicates(tag.name);

  if (includes(removedDuplicates, tag.name)) {
    logger.warn(`${tag.name} - removed as a duplicate`);
    await scrapeAlbums();
    return;
  }
  await scrapeAlbumsByTag(tag);
  await database.query(SQL`
    UPDATE "Tag"
    SET "albumsScrapedAt" = NOW()
    WHERE "name" = ${tag.name}
  `);
  logger.debug(`scrapeAlbums: ${tag.name} - success`);
  await logToTelegram(
    `\\#scrape\\_albums\nУспішно зібрано альбоми для тега *${escapeTelegramMessage(
      tag.name,
    )}*`,
  );
}
